FScriptTE.pas

```diff
@@ -0,0 +1,825 @@
+    form14.Label1.Hide;
+    form14.Show;
+    form14.ProgressBar1.max := Form4.ListBox1.items.count - 1;
+    for i := 0 to Form4.ListBox1.items.count - 1 do
+    begin
+      form14.ProgressBar1.Position := i;
+      form14.Repaint;
+      TextEdit.Lines.Add(Form4.ListBox1.items[i]);
+    end;
+    TextEdit.MoveCaretToBeginning;
+    form14.Hide;
+    form14.Caption := '3D Processing';
+    form14.ProgressBar1.Position := 1;
+    form14.Label1.Show;
+end;
+
+procedure TfmScriptTE.GoToLabel1Click(Sender: TObject);
+begin
+  fmGoto.ShowModal;
+end;
+
+procedure TfmScriptTE.Hex1Click(Sender: TObject);
+begin
+  form4.Hex1Click(nil);
+end;
+
+procedure TfmScriptTE.Newlabel1Click(Sender: TObject);
+var
+  i, j, lastline, lastcaret, labellength: integer;
+  found: Boolean;
+  whitespace: string;
+begin
+  lastline := TextEdit.TextPosition.Line;
+  lastcaret := TextEdit.CaretIndex;
+  whitespace := '';
+  // Add the next unused label
+  with TextEdit do
+  begin
+    for i := 0 to 65535 do
+    begin
+      found := false;
+      for j := 0 to Lines.Count do
+      begin
+        if Lines[j].StartsWith(inttostr(i) + ':') then
+        found := true;
+      end;
+      if not found then
+      begin
+        labellength := 6 - length(inttostr(i));
+        for j := 0 to labellength do
+          whitespace := whitespace + ' ';
+        InsertLine(lastline + 2, inttostr(i) + ':' + whitespace);
+        GoToLine(lastline + 1);
+        // Move caret to end of the new line
+        CaretIndex := lastcaret + 1 + length(Lines[lastline + 1]);
+        break;
+      end;
+    end;
+  end;
+end;
+
+procedure TfmScriptTE.Newregister1Click(Sender: TObject);
+var
+  i, j: integer;
+  found: Boolean;
+begin
+  // Add the next unused register
+  with TextEdit do
+  begin
+    for i := 0 to 255 do
+    begin
+      found := false;
+      for j := 0 to Lines.Count do
+      begin
+        if Lines[j].Contains('R' + inttostr(i)) then
+          found := true;
+      end;
+      if not found
+      // Exclude all reserved registers
+      and (i <> 74) and (i <> 75)
+      and (i <> 76) and (i <> 77)
+      and (i <> 78) and (i <> 79)
+      and (i <> 253) and (i <> 255)
+      then
+      begin
+        InsertText('R' + inttostr(i));
+        break;
+      end;
+    end;
+  end;
+end;
+
+procedure TfmScriptTE.Openfromfile1Click(Sender: TObject);
+begin
+  if opendialog1.Execute then
+  begin
+    Textedit.LoadFromFile(opendialog1.FileName);
+    textedited:=true;
+    isedited:=true;
+  end;
+end;
+
+procedure TfmScriptTE.Paste1Click(Sender: TObject);
+begin
+  TextEdit.PasteFromClipboard;
+end;
+
+procedure TfmScriptTE.Replace1Click(Sender: TObject);
+begin
+  fmReplace.ShowModal;
+end;
+
+procedure TfmScriptTE.Savetofile1Click(Sender: TObject);
+begin
+  if savedialog1.Execute then
+  begin
+    Textedit.SaveToFile(savedialog1.FileName);
+    isedited:=true;
+  end;
+end;
+
+procedure TfmScriptTE.TextEditChange(Sender: TObject);
+var
+  l,x,y,g: integer;
```

Comment:
this might make the edit look sluggish on some computers.
Maybe find a way to scan only on when caret change line?